ui/desktop/electron-app/src/session/index.js

```diff
@@ -0,0 +1,111 @@
+const { spawnAsyncJSONPromise } = require('../helpers/spawn-promise.js');
+/**
+ * Super paranoid shell quote/escape and validation.  Input must be base62.
+ * @param {string} str
+ */
+const escapeAndValidateBase62 = (str) => {
+  const candidate = str.toString();
+  if (candidate.match(/^[A-Za-z0-9_]*$/)) return candidate;
+  throw new Error(`
+    Could not invoke command:
+    input contained unsafe characters.
+  `);
+};
+
+class Session {
+  #id;
+  #addr;
+  #token;
+  #hostId;
+  #process;
+  #targetId;
+  #proxyDetails;
+
+  /**
+   * Initialize a session to a controller address
+   * using target details.
+   * @param {string} addr
+   * @param {string} targetId
+   * @param {string} token
+   * @param {string} hostId
+   */
+  constructor(addr, targetId, token, hostId) {
+    this.#addr = addr;
+    this.#targetId = targetId;
+    this.#token = token;
+    this.#hostId = hostId;
+  }
+
+  /**
+   * Session id
+   * @return {string}
+   */
+  get id() {
+    return this.#id;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  get isActive() {
+    return !this.#process.killed;
+  }
+
+  /**
+   * Using cli, initialize a session to a target.
+   * Tracks local proxy details if successful.
+   */
+  start() {
+    const command = this.cliCommand();
+    return spawnAsyncJSONPromise(command).then((spawnedSession) => {
+      this.#process = spawnedSession.childProcess;
+      this.#proxyDetails = spawnedSession.response;
+      this.#process = spawnedSession.childProcess;
+      this.#id = this.#proxyDetails.session_id;
+      return this.#proxyDetails;
+    });
+  }
+
+  /**
+   * Stop proxy process used by session.
+   */
+  stop() {
+    return new Promise((resolve, reject) => {
+      if (this.isActive) {
+        this.#process.on('close', () => resolve());
+        this.#process.on('error', (e) => reject(e));
+        this.#process.kill();
+      } else {
+        // Do nothing when process isn't active
+        resolve();
+      }
+    });
+  }
+
+  /**
+   * Generate cli command for session.
+   * @returns {[]}
+   */
+  cliCommand() {
+    const sanitized = {
+      target_id: escapeAndValidateBase62(this.#targetId),
+      token: escapeAndValidateBase62(this.#token),
+    };
+
+    const command = [
+      'connect',
+      `-target-id=${sanitized.target_id}`,
+      `-token=${sanitized.token}`,
+      `-addr=${this.#addr}`,
```

Comment:
This unsanitized input is bugging me a little.  We know it's highly unlikely to be unsafe since we validate the origin.  But considering the code in isolation, this feels a bit unsafe.  It can't use the same base62 scheme but is there a quick and dirty way to improve this situation?

Reply:
Attempted to use `encodeURI` to mimic sanitization. Any better alternatives?

Reply:
Not sure.  We need to prevent injection attacks and `encodeURI` allows most punctuation characters through, including semicolons.  This could still be used for script injection.  https://auth0.com/blog/preventing-command-injection-attacks-in-node-js-apps/

Reply:
Very interesting point here. I will do some research about this too. In case we can't find something "default" create a helper function to sanitize URI's within `core` addon (so we can share it between admin and desktop) is a possibility?

Reply:
I'm not sure there's much ability to share code between electron (Node) and non-electron (Ember) code.  The article above cites this tool:  https://www.npmjs.com/package/shell-quote for escaping shell inputs.